Doc/reference/datamodel.rst

```diff
@@ -889,57 +891,229 @@ Attribute assignment updates the module's namespace dictionary, e.g.,
+
+   The :term:`package` a module belongs to.
+
+   If the module is top-level (i.e. not a part of any specific package)
+   then the attribute should be set to ``''``. Otherwise, it should be set to
+   the name of the module's package (which can be equal to
+   :attr:`module.__name__` if the module itself is a package). See :pep:`366`
+   for further details.
+
+   This attribute is used instead of :attr:`~module.__name__` to calculate
+   explicit relative imports for main modules. It defaults to ``None`` for
+   modules created dynamically using the :class:`types.ModuleType` constructor;
+   use :func:`importlib.util.module_from_spec` instead to ensure the attribute
+   is set to a :class:`str`.
+
+   It is **strongly** recommended that you use
+   :attr:`module.__spec__.package <importlib.machinery.ModuleSpec.parent>`
+   instead of :attr:`!module.__package__`.
+
+   .. versionchanged:: 3.4
+      This attribute now defaults to ``None`` for modules created dynamically
+      using the :class:`types.ModuleType` constructor.
+      Previously the attribute was optional.
+
+   .. versionchanged:: 3.6
+      The value of :attr:`!__package__`` is expected to be the same as
+      :attr:`__spec__.parent <importlib.machinery.ModuleSpec.parent>`.
+
+   .. versionchanged:: 3.10
+      :exc:`ImportWarning` is raised if import falls back to
+      :attr:`!__package__` instead of
+      :attr:`~importlib.machinery.ModuleSpec.parent`.
+
+   .. versionchanged:: 3.12
+      Raise :exc:`DeprecationWarning` instead of :exc:`ImportWarning`
+      when falling back to :attr:`!__package__`.
+
+   .. deprecated-removed:: 3.13 3.15
+      :attr:`!__package__` will cease to be set or taken into consideration
+      by the import system or standard library.
+
+.. attribute:: module.__loader__
+
+   The :term:`loader` object that the import machinery used to load the module.
+
+   This attribute is mostly useful for introspection, but can be used for
+   additional loader-specific functionality, for example getting data
+   associated with a loader.
+
+   :attr:`!__loader__` defaults to ``None`` for modules created dynamically
+   using the :class:`types.ModuleType` constructor;
+   use :func:`importlib.util.module_from_spec` instead to ensure the attribute
+   is set to a :term:`loader` object.
+
+   It is **strongly** recommended that use
+   :attr:`module.__spec__.loader <importlib.machinery.ModuleSpec.loader>`
+   instead of :attr:`!module.__loader__`.
+
+   .. versionchanged:: 3.4
+      This attribute now defaults to ``None`` for modules created dynamically
+      using the :class:`types.ModuleType` constructor.
+      Previously the attribute was optional.
+
+   .. versionchanged:: 3.12
+      The value of :attr:`!module__loader__` is expected to be the same as
+      :attr:`__spec__.loader <importlib.machinery.ModuleSpec.loader>`.
+      The use of :attr:`!module.__loader__` is deprecated and slated for
+      removal in Python 3.14.
```

Comment:
Should this use a `deprecated-removed`?

Is it documented in What's New and `deprecations/pending-removal-in-3.14.rst`?

Reply:
> Should this use a `deprecated-removed`?

I think that makes sense!

> Is it documented in What's New and `deprecations/pending-removal-in-3.14.rst`?

It doesn't appear so. This note is mostly copied from the note that currently exists at https://docs.python.org/3/reference/import.html#loader__, but with some more cross-reference links added.

I'll change the directive to be `.. deprecated-removed`, but I'm _kinda_ tempted to say that adding a note to "What's new" is out of scope for this PR, which essentially moves documentation around and consolidates it rather than adding new documentation.

Reply:
I took a stab at improving this a bit in https://github.com/python/cpython/pull/124709/commits/309a9cf03d386b5467cddc8a1de7da4e080b7fc8 and https://github.com/python/cpython/pull/124709/commits/2dca569262d7dd215e6276ad964f1145db47e594!